upbc/protoc-gen-upb.cc

```diff
@@ -843,139 +843,163 @@ std::vector<TableEntry> FastDecodeTable(const protobuf::Descriptor* message,
   return table;
 }
 
-void WriteSource(const protobuf::FileDescriptor* file, Output& output,
-                 bool fasttable_enabled) {
-  EmitFileWarning(file, output);
+void WriteField(const protobuf::FieldDescriptor* field,
+                absl::string_view offset, absl::string_view presence,
+                int submsg_index, Output& output) {
 
-  output(
-      "#include <stddef.h>\n"
-      "#include \"upb/msg_internal.h\"\n"
-      "#include \"$0\"\n",
-      HeaderFilename(file->name()));
+  std::string label;
+  if (field->is_map()) {
+    label = "_UPB_LABEL_MAP";
+  } else if (field->is_packed()) {
+    label = "_UPB_LABEL_PACKED";
+  } else {
+    label = absl::StrCat(field->label());
+  }
 
-  for (int i = 0; i < file->dependency_count(); i++) {
-    output("#include \"$0\"\n", HeaderFilename(file->dependency(i)->name()));
+  output("{$0, $1, $2, $3, $4, $5}", field->number(), offset, presence,
+         submsg_index, TableDescriptorType(field), label);
+}
+
+// Writes a single field into a .upb.c source file.
+void WriteMessageField(const protobuf::FieldDescriptor* field,
+                       const MessageLayout& layout, int submsg_index,
+                       Output& output) {
+  std::string presence = "0";
+
+  if (MessageLayout::HasHasbit(field)) {
+    int index = layout.GetHasbitIndex(field);
+    assert(index != 0);
+    presence = absl::StrCat(index);
+  } else if (field->real_containing_oneof()) {
+    MessageLayout::Size case_offset =
+        layout.GetOneofCaseOffset(field->real_containing_oneof());
+
+    // We encode as negative to distinguish from hasbits.
+    case_offset.size32 = ~case_offset.size32;
+    case_offset.size64 = ~case_offset.size64;
+    assert(case_offset.size32 < 0);
+    assert(case_offset.size64 < 0);
+    presence = GetSizeInit(case_offset);
   }
 
-  output(
-      "\n"
-      "#include \"upb/port_def.inc\"\n"
-      "\n");
+  output("  ");
+  WriteField(field, GetSizeInit(layout.GetFieldOffset(field)), presence,
+             submsg_index, output);
+  output(",\n");
+}
 
+// Writes a single message into a .upb.c source file.
+void WriteMessage(const protobuf::Descriptor* message, Output& output,
+                  bool fasttable_enabled) {
```

Comment:
fast_table_enabled

Reply:
I would prefer to keep this, it matches:

- https://github.com/protocolbuffers/upb/blob/master/BUILD#L51-L60
- https://github.com/protocolbuffers/upb/blob/master/upb/msg_internal.h#L58
- https://github.com/protocolbuffers/upb/blob/master/upb/msg_internal.h#L72

It sets it apart as a specific term of art in the code. I think `minitable` would be a good follow-up term of art for mini tables.